### PR TITLE
php: builder

### PR DESCRIPTION
**Pull request by tinybeachthor**
Towards #240 

- [x] Add a builder
- [x] Add php example
- [x] dev shell
- [x] Update php docs
- [x] more robust semver implementation
- [x] discoverer

Disable packagist.org repo for composer, use [artifact](https://getcomposer.org/doc/05-repositories.md#artifact) or path repositories.

### CODE REVIEW

**Comment by tinybeachthor** [comment]
Tested on https://github.com/composer/composer, build and devShell both work as expected.

File: src/subsystems/php/discoverers/default/default.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  subsystem,
+  ...
+}: let
+  l = lib // builtins;
+
+  # get translators for the project
+  getTranslators = path: let
+    nodes = l.readDir path;
+  in
+    l.optionals (nodes ? "composer.lock") ["composer-lock"]
```
Comment (leungbk) [inline src/subsystems/php/discoverers/default/default.nix:13]:
```suggestion
    l.optional (nodes ? "composer.lock") "composer-lock"
```

**Comment by DavHau** [comment]
> * [ ]  ~Add php examples~ (need the builder merged first)

That should not be necessary. You can add the example flake under /examples/php/flake.nix, and then execute from the repo root: `nix run .#tests-examples php`

Alternatively navigate to /examples/php, then run:
`nix flake lock --override-input dream2nix ../../`
and
`nix run .#resolveImpre` / `nix build` ..etc.

File: src/subsystems/php/discoverers/default/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  dlib,
+  lib,
+  subsystem,
+  ...
+}: let
+  l = lib // builtins;
+
+  # get translators for the project
+  getTranslators = path: let
+    nodes = l.readDir path;
```
Comment (DavHau) [inline src/subsystems/php/discoverers/default/default.nix:11]: call to readDir could can be prevented by accessing `tree`.
Example: `tree.files ? "composer.lock"`

File: src/subsystems/php/translators/composer-lock/default.nix [hunk 3]
```diff
@@ -246,15 +245,15 @@ in {
       };
 
       # name of the default package
-      defaultPackage = composerJson.name;
+      defaultPackage = composerJson.name or "unknown";
```
Comment (DavHau) [inline src/subsystems/php/translators/composer-lock/default.nix:248]: The `name` is already determined by the discoverer. I think we should just re-use `project.name` here.

**Comment by DavHau** [comment]
I just added an example via e77c129ad6643cf7deb26d71e00c551e89ab285b. Working fine.

**Comment by DavHau** [comment]
It seems like the composer example adds significant overhead to the test execution. That is probably because of the slow nix git fetching. Maybe we can replace the example with a project, that doesn't have many dependencies.

File: src/subsystems/php/discoverers/default/default.nix (hunk 2 above)
Comment (tinybeachthor) [reply src/subsystems/php/discoverers/default/default.nix]: done

File: src/subsystems/php/translators/composer-lock/default.nix (hunk 3 above)
Comment (tinybeachthor) [reply src/subsystems/php/translators/composer-lock/default.nix]: done

File: src/subsystems/php/builders/simple/default.nix [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      allDependencies = let
+        getAllDependencies = deps: let
+          getSubdependencies = dep: let
+            subdeps = getDependencies dep.name dep.version;
+          in
+            getAllDependencies subdeps;
+        in
+          deps ++ (l.flatten (map getSubdependencies deps));
+      in
+        getAllDependencies dependencies;
```
[the remark this answers is not in the document]
Comment (tinybeachthor) [reply src/subsystems/php/builders/simple/default.nix]: nice, didn't know about `genericClosure`

**Comment by tinybeachthor** [comment]
@DavHau 
I've disabled checks on the composer example, which speeds it up a little.
Running the tests locally I get:
```
Execution times:
1 sec: _d2n-indexer
37 sec: php_composer        <<--
44 sec: nodejs_workspace
52 sec: _d2n-extended-new-subsystem
56 sec: _d2n-init-pkgs
56 sec: rust_ripgrep
83 sec: nodejs_prettier
85 sec: _d2n-extended
131 sec: nodejs_eslint
143 sec: debian-binary
144 sec: rust_set-rust-toolchain
155 sec: rust_no-cargo-lock
163 sec: nodejs_mattermost-webapp
175 sec: python_aiohttp
209 sec: haskell_stack-lock
210 sec: nodejs_eslint-aggreagted
281 sec: haskell_cabal-plan
```

**Comment by aszenz** [comment]
Thanks for this, I recently tried this builder on one of the projects but got this error:

```
error: Constraint "as" could not be parsed
```
The project composer.json uses aliases as described here: 
https://getcomposer.org/doc/articles/aliases.md#aliases

**Comment by tinybeachthor** [comment]
@aszenz 
The builder does not handle some less-common composer features at the moment (like composer plugins, php version constraints, composer api, aliases, ...).
Could you link to the project or if it's a private, make a small example of the composer.json that produces this error?

**Comment by aszenz** [comment]
> @aszenz Thank you for noticing and reporting this. The builder does not handle some less-common composer features at the moment (like composer plugins, php version constraints, composer api, aliases, update/install script hooks, ...). Could you link to the project or if it's a private, make a small example of the composer.json that produces this error?

Sure here's a sample project on which running `nix build .#packages.x86_64-linux.app/php` fails with:
`error: Constraint "as" could not be parsed`

https://github.com/aszenz/php-app-nix

If i remove the `as` then i get an error about missing extensions, seems like the php extensions mentioned in `composer.json` are not available when the build runs

**Comment by tinybeachthor** [comment]
@aszenz 
#296 fixes both issues (inline version alias, extra php extenstions)

**Pull request merged**